Sources/SwiftFusion/Inference/ValuesStorage.swift

```diff
@@ -17,368 +17,369 @@
-  var tangentIdentifier: ObjectIdentifier {
-    storage.tangentIdentifier
+// MARK: - Algorithms on arrays of `EuclideanVectorN` values.
+
+extension ArrayStorage where Element: EuclideanVectorN {
+  /// Adds `others` to `self`.
+  ///
+  /// - Requires: `others.count >= count`.
+  func add(_ others: ArrayBuffer<Element>) {
+    withUnsafeMutableBufferPointer { vs in
+      others.withUnsafeBufferPointer { os in
+        for i in vs.indices {
+          vs[i] += os[i]
+        }
+      }
+    }
   }
 
-  /// The values' zero `TangentVector`s.
-  var zeroTangent: AnyArrayBuffer<AnyVectorStorage> {
-    storage.zeroTangent
+  /// Scales each element of `self` by `scalar`.
+  func scale(by scalar: Double) {
+    withUnsafeMutableBufferPointer { vs in
+      for i in vs.indices {
+        vs[i] *= scalar
+      }
+    }
   }
 
-  /// Moves each element of `self` along the corresponding element of `directions`.
+  /// Returns the dot product of `self` with `others`.
   ///
-  /// Precondition: `direction` has at least `count` elements of type `Element.TangentVector`,
-  /// where `Element` is the type of `self`.
-  mutating func move<DirectionStorage>(along directions: AnyArrayBuffer<DirectionStorage>) {
-    ensureUniqueStorage()
-    directions.withUnsafeRawPointerToElements { directionsStart in
-      storage.move(along: directionsStart)
+  /// This is the sum of the dot products of corresponding elements.
+  ///
+  /// - Requires: `others.count >= count`.
+  func dot(_ others: ArrayBuffer<Element>) -> Double {
+    withUnsafeMutableBufferPointer { vs in
+      others.withUnsafeBufferPointer { os in
+        vs.indices.reduce(into: 0) { (result, i) in result += vs[i].dot(os[i]) }
+      }
     }
   }
+
+  /// For each variable, a Jacobian that scales it by `scalar`.
+  func jacobians(scalar: Double) -> AnyGaussianFactorArrayBuffer {
+    AnyGaussianFactorArrayBuffer(ArrayBuffer(enumerated().lazy.map { (i, _) in
+      ScalarJacobianFactor(edges: Tuple1(TypedID<Element, Int>(i)), scalar: scalar)
+    }))
+  }
 }
 
-/// Contiguous storage of homogeneous `Differentiable` values of statically unknown type.
-protocol AnyDifferentiableStorageImplementation: AnyDifferentiableStorage {
-  /// Identifies the values' `TangentVector`.
-  var tangentIdentifier_: ObjectIdentifier { get }
+// MARK: - Type-erased arrays of `Differentiable` values.
 
-  /// The values' zero `TangentVector`s.
-  var zeroTangent_: AnyArrayBuffer<AnyVectorStorage> { get }
+typealias AnyDifferentiableArrayBuffer = AnyArrayBuffer<DifferentiableArrayDispatch>
 
-  /// Moves `self` along the vector starting at `directionsStart`.
+/// An `AnyArrayBuffer` dispatcher that provides algorithm implementations for `Differentiable`
+/// elements.
+class DifferentiableArrayDispatch: AnyObjectDispatch {
+  /// The `TangentVector` type of the elements.
+  class var tangentVectorType: Any.Type {
+    fatalError("implement as in DifferentiableArrayDispatch_")
+  }
+
```

Comment:
It is used in `VariableAssignments`'s `tangentVectorZeros` to get the dictionary key for the dictionary of tangent vectors. (When we switch to the new TypedID thing that lets you have custom keys we won't need this any more because then the key will be the type of the original variable that it comes from.)

I'll add a test for now.

Reply:
actually, there already is a test